File: app/src/main/java/com/kota/Telnet/Reference/TelnetAnsiCode.java
package com.kota.Telnet.Reference;

import android.annotation.SuppressLint;
import android.view.View;

import androidx.core.internal.view.SupportMenu;
import androidx.core.view.InputDeviceCompat;

import com.kota.Telnet.TelnetAnsi;

public class TelnetAnsiCode {
    public static final int[] BACKGROUND_COLOR_NORMAL = {View.MEASURED_STATE_MASK, -8388608, -16744448, -8355840, -16777088, -8388480, -16744320, -4144960};
    @SuppressLint("RestrictedApi")
    public static final int[] COLOR_BRIGHT = {-8355712, SupportMenu.CATEGORY_MASK, -16711936, InputDeviceCompat.SOURCE_ANY, -16776961, -65281, -16711681, -1};
    public static final int[] TEXT_COLOR_NORMAL = {View.MEASURED_STATE_MASK, -8388608, -16744448, -8355840, -16777088, -8388480, -16744320, -4144960};

    public static class Code {
        public static final int CHA = 6;
        public static final int CNL = 4;
        public static final int CPL = 5;
        public static final int CUB = 3;
        public static final int CUD = 1;
        public static final int CUF = 2;
        public static final int CUP = 7;
        public static final int CUU = 0;
        public static final int DSR = 14;
        public static final int ED = 8;
        public static final int EL = 9;
        public static final int HC = 17;
        public static final int HVP = 12;
        public static final int RCP = 16;
        public static final int SC = 18;
        public static final int SCP = 15;
        public static final int SD = 11;
        public static final int SGR = 13;
        public static final int SU = 10;
    }

    public static class Color {
        public static final byte BLACK = 0;
        public static final byte RED = 1;
        public static final byte GREEN = 2;
        public static final byte YELLOW = 3;
        public static final byte BLUE = 4;
        public static final byte MAGENTA = 5;
        public static final byte CYAN = 6;
        public static final byte WHITE = 7;
    }

    /**
     * 瀏覽文章, 返回前景色
     * @param colorIndex index
     * @return int (color code)
     */
    public static int getTextColor(byte colorIndex) {
        int color_index = colorIndex & 255;
        if (colorIndex < 8) {
            return TEXT_COLOR_NORMAL[color_index];
        }
        try {
            return COLOR_BRIGHT[color_index - 8];
        } catch (Exception e) {
            e.printStackTrace();
            return -4144960;
        }
    }

    /**
     * 瀏覽文章, 返回背景色
     * @param colorIndex index
     * @return int (color code)
     */
    public static int getBackgroundColor(byte colorIndex) {
        int color_index = colorIndex & 255;
        if (colorIndex < 8) {
            return BACKGROUND_COLOR_NORMAL[color_index];
        }
        try {
            return COLOR_BRIGHT[color_index - 8];
        } catch (Exception e) {
            e.printStackTrace();
            return -4144960;
        }
    }

    /**
     * 修改文章, 返回前景字碼
     * @param colorIndex  index
     * @return string
     */
    public static String getTextAsciiCode(byte colorIndex) {
        int color_index = colorIndex & 255;
        if (color_index == TelnetAnsi.getDefaultTextColor()) {
            return "";
        }
        if (colorIndex < 8) {
            return "3" + colorIndex;
        }
        try {
            return "1;3" + (color_index - 8);
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    /**
     * 修改文章, 返回背景字碼
     * @param colorIndex  index
     * @return string
     */
    public static String getBackAsciiCode(byte colorIndex) {
        int color_index = colorIndex & 255;
        if (color_index == TelnetAnsi.getDefaultBackgroundColor()) {
            return "";
        }
        if (colorIndex < 8) {
            return "4" + colorIndex;
        }
        return "";
    }
}

File: app/src/main/java/com/kota/Telnet/Reference/TelnetAnsiCode.kt
package com.kota.Telnet.Reference

import android.annotation.SuppressLint
import android.util.Log
import android.view.View
import androidx.core.internal.view.SupportMenu
import androidx.core.view.InputDeviceCompat

object TelnetAnsiCode {
    val BACKGROUND_COLOR_NORMAL = intArrayOf(
        View.MEASURED_STATE_MASK, -8388608, -16744448, -8355840, 
        -16777088, -8388480, -16744320, -4144960
    )
    
    @SuppressLint("RestrictedApi")
    val COLOR_BRIGHT = intArrayOf(
        -8355712, SupportMenu.CATEGORY_MASK, -16711936, InputDeviceCompat.SOURCE_ANY, 
        -16776961, -65281, -16711681, -1
    )
    
    val TEXT_COLOR_NORMAL = intArrayOf(
        View.MEASURED_STATE_MASK, -8388608, -16744448, -8355840, 
        -16777088, -8388480, -16744320, -4144960
    )

    object Code {
        const val CHA = 6
        const val CNL = 4
        const val CPL = 5
        const val CUB = 3
        const val CUD = 1
        const val CUF = 2
        const val CUP = 7
        const val CUU = 0
        const val DSR = 14
        const val ED = 8
        const val EL = 9
        const val HC = 17
        const val HVP = 12
        const val RCP = 16
        const val SC = 18
        const val SCP = 15
        const val SD = 11
        const val SGR = 13
        const val SU = 10
    }

    object Color {
        const val BLACK: Byte = 0
        const val RED: Byte = 1
        const val GREEN: Byte = 2
        const val YELLOW: Byte = 3
        const val BLUE: Byte = 4
        const val MAGENTA: Byte = 5
        const val CYAN: Byte = 6
        const val WHITE: Byte = 7
    }

    /**
     * 瀏覽文章, 返回前景色
     * @param colorIndex index
     * @return Int (color code)
     */
    fun getTextColor(colorIndex: Byte): Int {
        val colorIndexInt = colorIndex.toInt() and 255
        return if (colorIndex < 8) {
            TEXT_COLOR_NORMAL[colorIndexInt]
        } else {
            try {
                COLOR_BRIGHT[colorIndexInt - 8]
            } catch (e: Exception) {
                Log.e(TelnetAnsiCode::class.java.simpleName, e.message ?: "")
                -4144960
            }
        }
    }

    /**
     * 瀏覽文章, 返回背景色
     * @param colorIndex index
     * @return Int (color code)
     */
    fun getBackgroundColor(colorIndex: Byte): Int {
        val colorIndexInt = colorIndex.toInt() and 255
        return if (colorIndex < 8) {
            BACKGROUND_COLOR_NORMAL[colorIndexInt]
        } else {
            try {
                COLOR_BRIGHT[colorIndexInt - 8]
            } catch (e: Exception) {
                Log.e(TelnetAnsiCode::class.java.simpleName, e.message ?: "")
                -4144960
            }
        }
    }

    /**
     * 修改文章, 返回前景字碼, 亮色已經先處理掉
     * @param colorIndex index
     * @return String
     */
    fun getTextAsciiCode(colorIndex: Byte): String {
        return if (colorIndex < 8) {
            "3$colorIndex"
        } else {
            ""
        }
    }

    /**
     * 修改文章, 返回背景字碼
     * @param colorIndex index
     * @return String
     */
    fun getBackAsciiCode(colorIndex: Byte): String {
        return if (colorIndex < 8) {
            "4$colorIndex"
        } else {
            ""
        }
    }
}
